### PR TITLE
Product

### PR DESCRIPTION
**Pull request by LorDEvilDead**
Короче, я намешал в контроллере много чего и теперь не понимаю что за чем идёт и почему, есть пара вопросв.
1-й вопрос, двигаюсь я в правильном направлении или нет?

### CODE REVIEW

**Comment by MikhailZaitsau** [comment]
FIrst things first, you should ALWAYS named your commits clearly and informative

File: backend/db/migrate/20240502191526_create_products.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class CreateProducts < ActiveRecord::Migration[7.1]
+  def change
+    create_table :products do |t|
+      t.decimal :price
+      t.integer :weight
+      t.string :consist
```
Comment (MikhailZaitsau) [inline backend/db/migrate/20240502191526_create_products.rb]: `ingredients`?

File: backend/spec/factories/products.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :product do
+    price { "9.99" }
```
Comment (MikhailZaitsau) [inline backend/spec/factories/products.rb]: decimal?

File: backend/db/migrate/20240502191526_create_products.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      t.text :disription
```
Comment (MikhailZaitsau) [inline backend/db/migrate/20240502191526_create_products.rb]: `description`?

File: backend/db/migrate/20240502191526_create_products.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class CreateProducts < ActiveRecord::Migration[7.1]
+  def change
+    create_table :products do |t|
+      t.decimal :price
+      t.integer :weight
```
Comment (MikhailZaitsau) [inline backend/db/migrate/20240502191526_create_products.rb]: `portion_weight_grams`?

File: backend/db/migrate/20240502191526_create_products.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      t.string :consist
+      t.text :disription
+      t.string :photo
```
Comment (MikhailZaitsau) [inline backend/db/migrate/20240502191526_create_products.rb]: `image`?

File: backend/spec/factories/products.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+    weight { 1 }
+    consist { "MyString" }
+    disription { "MyText" }
+    photo { "MyString" }
+    name { "MyString" }
```
Comment (MikhailZaitsau) [inline backend/spec/factories/products.rb]: Explore [Faker ](https://github.com/faker-ruby/faker), it is cool and usefull.
Try `Faker::Lorem.word` for names and `Faker::Lorem.paragraphs` for description.

File: backend/app/controllers/products_controller.rb
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb:1]: You always use same instance variables. Why? What about local variables?

File: backend/app/models/product.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+class Product < ApplicationRecord
+    validates :name, presence: true, lenght: {maximum: 100}
```
Comment (MikhailZaitsau) [inline backend/app/models/product.rb]: `"Блинчики из пшеничной муки с начинкой из ежевики с сахаром под сливочным соусом с посыпкой из сахорной пудры с корицей".length` == 118

File: backend/spec/factories/products.rb (hunk 6 above)
Comment (LorDEvilDead) [reply backend/spec/factories/products.rb]: > Explore [Faker ](https://github.com/faker-ruby/faker), it is cool and usefull. Try `Faker::Lorem.word` for names and `Faker::Lorem.paragraphs` for description.

@MikhailZaitsau Прикольно, для фабрик я писал ничего, от слова совсем %)

File: backend/db/migrate/20240502191526_create_products.rb (hunk 4 above)
Comment (LorDEvilDead) [reply backend/db/migrate/20240502191526_create_products.rb]: @MikhailZaitsau Хм, тут можно ещё подумать, потому что вес - понятно какой, но мб ты и прав

File: backend/spec/factories/products.rb (hunk 2 above)
Comment (LorDEvilDead) [reply backend/spec/factories/products.rb]: > decimal?

Я факторибота вообще не трогал)
Тут видимо сгенерилось по результатам моей генерации

File: .DS_Store
Comment (MikhailZaitsau) [inline .DS_Store:1]: Why this is here?

nobody: "Please, use `git add .` very careful... Yeah, fuck me..."
никто: «Пожалуйста, используй `git add .` с осторожностью... Ну-да, ну-да, пошёл я нахер..."

[This](https://stackoverflow.com/questions/7335420/global-git-ignore) and [this](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c) could be usefull for you if you want to add some file to `.gitignore` globaly

File: .env
Comment (MikhailZaitsau) [inline .env:1]: Why this is here?

File: backend/.DS_Store
Comment (MikhailZaitsau) [inline backend/.DS_Store:1]: Why this is here?

File: backend/app/controllers/products_controller.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    end
+    
+    #DELETE /products/1
+    def destroy
+    @product.destroy
+    end
+    
+    private
+    #not sure about that
+    attr_reader :params, :user
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: [Really clear explanation
](https://stackoverflow.com/questions/4370960/what-is-attr-accessor-in-ruby)

File: backend/app/controllers/products_controller.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        case product_creation
+        in Success(product)
+          render json: prepare_json(product), status: :created
+        in Failure(error)
+          render json: { errors: error.messages }, status: :unprocessable_entity
+        end
+    end
+
+    #PUT/PATCH /products/edit/1
+    def update
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: Will this method really update the product? Looks like 'set_product` do nothing except finding product.

File: backend/app/controllers/products_controller.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            render json: prepare_json(product), status :updated
+        in Failure
+            render json: { error: error.messages }, status :unprocessable_entity
+        end
+    end
+
+    #GET /products/1
+    def show
+        id = params.extract_value(:id)
+        @product = product.find(id)
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: What will happen if there is no record with requested id? And what do you gonna do with that?

File: backend/app/controllers/products_controller.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    end
+
+    #GET /products
+    def index
+        @products = Product.all
+        
+        render json @products
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: Shure you want to reder JSON with raw data from DB? Maybe it's better create hash with neccesary data before? Or use any [serializer](https://buttercms.com/blog/json-serialization-in-rails-a-complete-guide/)? Something like [RABL](https://github.com/nesquena/rabl) or [JSON API Resources](http://jsonapi-resources.com/)?

File: backend/app/controllers/products_controller.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        @product = Product.find(product_params[:id])
+    end
+
+    #call monades
+    def product_creation
+    Products::Create.new.call(params: product_params)
+    end
+
+     # Only allow a list of trusted parameters through.
+    def product_params
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: Not sure, but I think we need different params for create and update a record. What do you think about it? At least we definitely don’t need an id when creating.

File: backend/app/controllers/products_controller.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    attr_reader :params, :user
+
+    # Use callbacks to share common setup or constraints between actions.
+    def set_product
+        @product = Product.find(product_params[:id])
+    end
+
+    #call monades
+    def product_creation
+    Products::Create.new.call(params: product_params)
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: You pass `params` to the function, but it also waits for the `user`. It wouldn't work. We don't have authentication yet, so have no `current_user`, but you must use it in your code anyway.

File: backend/config/routes.rb [hunk 14]
```diff
@@ -7,7 +7,7 @@
   # Can be used by load balancers and uptime monitors to verify that the app is live.
   get 'up' => 'rails/health#show', as: :rails_health_check
   get 'index', to: 'messages#index'
-
+  resources :products
```
Comment (MikhailZaitsau) [inline backend/config/routes.rb:12]: Really need all possible resource routes for products now?

File: bin/.DS_Store
Comment (MikhailZaitsau) [inline bin/.DS_Store:1]: Why this is here?

File: docker/.DS_Store
Comment (MikhailZaitsau) [inline docker/.DS_Store:1]: Why this is here?

**Review by MikhailZaitsau — CHANGES_REQUESTED** [review]

File: .DS_Store
Comment (MikhailZaitsau) [inline .DS_Store:1]: Why this is here?

File: bin/.DS_Store
Comment (MikhailZaitsau) [inline bin/.DS_Store:1]: Why this is here?

File: docker/.DS_Store
Comment (MikhailZaitsau) [inline docker/.DS_Store:1]: Why this is here?

File: backend/app/controllers/products_controller.rb [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    end
+
+    #DELETE /products/1
+    def destroy
+        @product.destroy
+    end
+
+    private
+    #Use callbacks to share common setup or constrains between actions.
+    def set_product
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: What do this method? Why it is here?

File: backend/app/controllers/products_controller.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        @product = product.find(params[:id])
+    end
+
+    #only allow a list of trusted parameters though.
+    def group_params
+        params.require(:product).permit(:price, :portion_weight_grams, :ingredients, :desription, :image, :name)
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: Are you sure it is require a product instance to create or update a product instance?

File: backend/app/controllers/products_controller.rb [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+    #DELETE /products/1
+    def destroy
+        @product.destroy
+    end
+
+    private
+    #Use callbacks to share common setup or constrains between actions.
+    def set_product
+        @product = product.find(params[:id])
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: What will happen if there is no product with requested id in db? You must do something to handle it, do you?

File: backend/app/controllers/products_controller.rb [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    #GET /products/1
+    def show 
+        render json: @product
+    end
+
+    #POST /products
+    def create 
+        @product = Product.new(product_params)
+        if @product.save
+            render json: @gproduct, status: :created, location: @product
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: `@gproduct`? Looks like typo. And why did you add `location: @product`?

File: backend/app/controllers/products_controller.rb [hunk 19]
```diff
@@ -0,0 +1,48 @@
+class ProductsController < ApplicationController
+
+    #GET /products
+    def index
+        @products = Product.all
+        
+        render json: @products
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: Question about data preparing before json render is still actual

**Review by MikhailZaitsau — CHANGES_REQUESTED** [review]

File: backend/app/controllers/products_controller.rb (hunk 17 above)
Comment (LorDEvilDead) [reply backend/app/controllers/products_controller.rb]: Добавить исключения

File: backend/app/controllers/products_controller.rb [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    #GET /products
+    def index
+        products = Product.all
+        
+        render json: products.to_json(only: [:image, :price, :name, :portion_weight_grams, :ingredients])
+    end
+
+    #GET /products/1
+    def show 
+         render json: @product.to_json(only: [:price, :portion_weight_grams, :ingredients, :desription, :image, :name])
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: What will happen if there is no product with this id?

File: backend/spec/requests/products_spec.rb [hunk 21]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+# RSpec.describe 'Products' do
```
Comment (MikhailZaitsau) [inline backend/spec/requests/products_spec.rb]: Remove unnecessary comments

File: backend/app/controllers/products_controller.rb [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    def create
+        create_product(product_params)
+        @created_product ? (render json: @created_product, status: :created) : (head :bad_request)
+    end
+
+    def update
+        update_product(product_params)
+        @updated_product ? (render json: @updated_product, status: :updated) : (head :bad_request)
```
Comment (MikhailZaitsau) [inline backend/app/controllers/products_controller.rb]: `status: :updated`? What is this status code?

File: backend/app/controllers/users_controller.rb
Comment (MikhailZaitsau) [inline backend/app/controllers/users_controller.rb:1]: Excellently written controller, it is obvious that it was made by a pro. I really like it!

File: .env
Comment (MikhailZaitsau) [inline .env:1]: Why this is here?

File: backend/app/models/user.rb
Comment (MikhailZaitsau) [inline backend/app/models/user.rb:1]: Excellently written model, it is obvious that a pro made it. I like it very much!

File: backend/spec/requests/users_controller_spec.rb
Comment (MikhailZaitsau) [inline backend/spec/requests/users_controller_spec.rb:1]: Excellently written spec, it is obvious that a pro made it. I like it very much!

File: backend/app/services/product/update.rb [hunk 23]
```diff
@@ -0,0 +1,16 @@
+module Products
+    class Show < Core::Service
```
Comment (MikhailZaitsau) [inline backend/app/services/product/update.rb]: Looks like the class name is incorrect.

File: backend/app/models/product.rb [hunk 24]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class Product < ApplicationRecord
+  validates :name, presence: true, length: { maximum: 100 }
```
Comment (MikhailZaitsau) [inline backend/app/models/product.rb:4]: Don't you think we need additional validations? Product could not be created without price, ingredients and weight, could it?

Comment (LorDEvilDead) [reply backend/app/models/product.rb:4]: What about test product position or temporary position it one of case. What if ingridients will be secret? Or may be not wish public it? About price - what if product in pause to cook
@MikhailZaitsau

**Review by MikhailZaitsau — APPROVED** [review]
